File: src/main/java/com/cxxcxx/zinecraft/api/weapon/action/TimedWeaponActionRuntime.java
package com.cxxcxx.zinecraft.api.weapon.action;

import org.jetbrains.annotations.NotNull;

public abstract class TimedWeaponActionRuntime implements WeaponActionRuntime {
  @NotNull
  private final TickRange activeTicks;
  private final int durationTicks;
  private int currentTick;

  public TimedWeaponActionRuntime(@NotNull TickRange activeTicks, int durationTicks) {
    super();
    this.activeTicks = activeTicks;
    this.durationTicks = durationTicks;
    if (this.durationTicks <= 0) {
      int j = 0;
      String string1 = "动作持续时间必须大于 0";
      throw new IllegalArgumentException(string1.toString());
    }

    if (this.activeTicks.endInclusive() >= this.durationTicks) {
      int i = 0;
      String string = "生效阶段必须位于动作时间线内";
      throw new IllegalArgumentException(string.toString());
    }
  }

  @Override
  public final int getCurrentTick() {
    return this.currentTick;
  }

  @NotNull
  @Override
  public final ActionPhase getPhase() {
    ActionPhase actionPhase;
    if (this.currentTick < this.activeTicks.startInclusive()) {
      actionPhase = ActionPhase.STARTUP;
    } else {
      actionPhase = this.activeTicks.contains(this.currentTick)
          ? ActionPhase.ACTIVE
          : (this.currentTick < this.durationTicks ? ActionPhase.RECOVERY : ActionPhase.FINISHED);
    }

    return actionPhase;
  }

  @Override
  public final boolean getFinished() {
    return this.getPhase() == ActionPhase.FINISHED;
  }

  @Override
  public final void tick() {
    if (!this.getFinished()) {
      this.onTick(this.currentTick);
      int i = this.currentTick++;
    }
  }

  protected abstract void onTick(int var1);
}

File: src/main/kotlin/com/cxxcxx/zinecraft/api/weapon/action/TimedWeaponActionRuntime.kt
package com.cxxcxx.zinecraft.api.weapon.action

/** 用于只有单段启动/生效/后摇的动作，避免每种武器重复 tick 状态机。 */
abstract class TimedWeaponActionRuntime(
  private val activeTicks: IntRange,
  private val durationTicks: Int
) : WeaponActionRuntime {
  init {
    require(durationTicks > 0) { "动作持续时间必须大于 0" }
    require(activeTicks.first >= 0 && activeTicks.last < durationTicks) { "生效阶段必须位于动作时间线内" }
  }

  final override var currentTick: Int = 0
    private set

  final override val phase: ActionPhase
    get() = when {
      currentTick < activeTicks.first -> ActionPhase.STARTUP
      currentTick in activeTicks -> ActionPhase.ACTIVE
      currentTick < durationTicks -> ActionPhase.RECOVERY
      else -> ActionPhase.FINISHED
    }

  final override val finished: Boolean
    get() = phase == ActionPhase.FINISHED

  final override fun tick() {
    if (finished) return
    onTick(currentTick)
    currentTick++
  }

  protected abstract fun onTick(tick: Int)
}
